stan/math/opencl/matrix_cl.hpp

```diff
@@ -39,34 +36,104 @@ class matrix_cl {
   cl::Buffer oclBuffer_;
   const int rows_;
   const int cols_;
+  mutable std::vector<cl::Event> write_events_;  // Tracks write jobs
+  mutable std::vector<cl::Event> read_events_;   // Tracks reads
 
  public:
+  // Forward declare the methods that work in place on the matrix
+  template <TriangularViewCL triangular_view = TriangularViewCL::Entire>
+  void zeros();
+  template <TriangularMapCL triangular_map = TriangularMapCL::LowerToUpper>
+  void triangular_transpose();
+  template <TriangularViewCL triangular_view = TriangularViewCL::Entire>
+  void sub_block(const matrix_cl& A, size_t A_i, size_t A_j, size_t this_i,
+                 size_t this_j, size_t nrows, size_t ncols);
   int rows() const { return rows_; }
 
   int cols() const { return cols_; }
 
   int size() const { return rows_ * cols_; }
 
-  const cl::Buffer& buffer() const { return oclBuffer_; }
+  /**
+   * Clear the write events from the event stacks.
+   */
+  inline void clear_write_events() const {
+    write_events_.clear();
+    return;
+  }
 
-  matrix_cl() : rows_(0), cols_(0) {}
+  /**
+   * Clear the read/write and write events from the event stacks.
+   */
+  inline void clear_read_events() const {
+    read_events_.clear();
+    return;
+  }
 
-  matrix_cl(const matrix_cl& A) : rows_(A.rows()), cols_(A.cols()) {
-    if (A.size() == 0)
-      return;
-    // the context is needed to create the buffer object
-    cl::Context& ctx = opencl_context.context();
-    try {
-      // creates a read&write object for "size" double values
-      // in the provided context
-      oclBuffer_ = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(double) * size());
+  /**
+   * Get the events from the event stacks.
+   * @return The write event stack.
+   */
+  inline const std::vector<cl::Event>& write_events() const {
+    return write_events_;
+  }
 
-      opencl_kernels::copy(cl::NDRange(rows_, cols_), A.buffer(),
-                           this->buffer(), rows_, cols_);
-    } catch (const cl::Error& e) {
-      check_opencl_error("copy (OpenCL)->(OpenCL)", e);
-    }
+  /**
+   * Get the events from the event stacks.
+   * @return The read/write event stack.
+   */
+  inline const std::vector<cl::Event>& read_events() const {
+    return read_events_;
   }
+
+  /**
+   * Add an event to the read/write event stack.
+   * @param new_event The event to be pushed on the event stack.
+   */
+  inline void add_read_event(cl::Event new_event) const {
+    this->read_events_.push_back(new_event);
```

Comment:
@seantalts I think we also want to add read events to the write stack here. The below will write before read error

```
A = B * C
D = A * 2
A = G * H
```

Right now we are only waiting for other write events before writes, but I think we also want to wait for all previous reads because of the above

Reply:
Yes sorry rereading my comment i was not being clear, pretty sure we are on the same page but to double check

```
// A waits for all reads and writes
// B C waits for previous writes
A = B * C
// D waits for all reads and writes
// A waits for all writes
D = A * 2
// A waits for all reads and writes
// G H waits for previous writes
A = G * H
```

So I think the rules here will be
1. A `read_buffer` only waits for all previous writers
2. A `write_buffer` waits for previous reads and writes

If you agree with the above then I'll fix this up. We can make a `read_write_event_stack` that gets called for tracking both reads and writes and is used when the matrix is declared a `write_buffer`

Reply:
Can we store the reads and writes independently w/o duplicating while providing the two methods we anticipate being used? Something like wait_all() and wait_writes() would make sense to me, with wait_all() concatting the read and write event vectors together.

Reply:
I added this, though now we have to have `select_events` return by value instead of reference. Because we do

```cpp
  inline const std::vector<cl::Event> read_write_events() const {
    return vec_concat(this->read_events(), this->write_events());
  }
```

We can't return this by reference, but I think if it's end result is also const then RVO kicks in and does the move for us? Though now we also have to do this

```cpp
  inline void wait_for_read_write_events() const {
...
    const std::vector<cl::Event> mat_events = this->read_write_events();
    queue.enqueueBarrierWithWaitList(&mat_events, &copy_event);
...
  }
```
Because doing
```cpp
  inline void wait_for_read_write_events() const {
...
    queue.enqueueBarrierWithWaitList(&this->read_write_events(), &copy_event);
...
  }
```
Tries taking the address of a temporary which is a no-no. 

I'm not really sure which one is best here. This method feels like it's a weird gotcha and I'd rather have the `read_write_` member in matrix_cl. But then of course we have to do the multiple assignments and have a weird issue when clearing out the read/write stacks individually but not clearing out the `read_write` stack in some situations

One very hacky way around this is to have a `read_write_stack` member in `matrix_cl` and assign/return it in `read_write_events`. Though I'm extremely not for that solution